controllers/eventhandlers/serviceimport.go

```diff
@@ -3,121 +3,47 @@ package eventhandlers
+	"github.com/aws/aws-application-networking-k8s/pkg/k8s"
 	mcs_api "sigs.k8s.io/mcs-api/pkg/apis/v1alpha1"
 )
 
-type enqueueRequestsForServiceImportEvent struct {
-	log       gwlog.Logger
-	client    client.Client
-	routeType core.RouteType
+type serviceImportEventHandler struct {
+	log    gwlog.Logger
+	client client.Client
+	mapper *resourceMapper
 }
 
-func NewEqueueRequestServiceImportEvent(log gwlog.Logger, client client.Client, routeType core.RouteType) handler.EventHandler {
-	return &enqueueRequestsForServiceImportEvent{
-		log:       log,
-		client:    client,
-		routeType: routeType,
+func NewServiceImportEventHandler(log gwlog.Logger, client client.Client) *serviceImportEventHandler {
+	return &serviceImportEventHandler{
+		log:    log,
+		client: client,
+		mapper: &resourceMapper{log: log, client: client},
 	}
 }
 
-func (h *enqueueRequestsForServiceImportEvent) Create(e event.CreateEvent, queue workqueue.RateLimitingInterface) {
-	newServiceImport := e.Object.(*mcs_api.ServiceImport)
-	h.enqueueImpactedService(queue, newServiceImport)
+func (h *serviceImportEventHandler) MapToRoute(routeType core.RouteType) handler.EventHandler {
+	return handler.EnqueueRequestsFromMapFunc(func(obj client.Object) []reconcile.Request {
+		return h.mapToRoute(obj, routeType)
+	})
 }
 
-func (h *enqueueRequestsForServiceImportEvent) Update(e event.UpdateEvent, queue workqueue.RateLimitingInterface) {
-	oldServiceImport := e.ObjectOld.(*mcs_api.ServiceImport)
-	newServiceImport := e.ObjectNew.(*mcs_api.ServiceImport)
-
-	if !equality.Semantic.DeepEqual(oldServiceImport, newServiceImport) {
-		h.enqueueImpactedService(queue, newServiceImport)
-	}
-}
-
-func (h *enqueueRequestsForServiceImportEvent) Delete(e event.DeleteEvent, queue workqueue.RateLimitingInterface) {
-	// TODO
-	oldServiceImport := e.Object.(*mcs_api.ServiceImport)
-	h.enqueueImpactedService(queue, oldServiceImport)
-
-}
-
-func (h *enqueueRequestsForServiceImportEvent) Generic(e event.GenericEvent, queue workqueue.RateLimitingInterface) {
-
-}
-
-func (h *enqueueRequestsForServiceImportEvent) enqueueImpactedService(queue workqueue.RateLimitingInterface, serviceImport *mcs_api.ServiceImport) {
-	h.log.Infof("enqueueImpactedRoute, serviceImport[%v]", serviceImport)
-
-	var routes []core.Route
-	var err error
-
-	switch h.routeType {
-	case core.HttpRouteType:
-		routes, err = core.ListHTTPRoutes(context.TODO(), h.client)
-	case core.GrpcRouteType:
-		routes, err = core.ListGRPCRoutes(context.TODO(), h.client)
-	default:
-		h.log.Errorf("Invalid routeType %s", h.routeType)
-		return
-	}
-
-	if err != nil {
-		h.log.Errorf("Error while listing routes, %s", err)
-	}
+func (h *serviceImportEventHandler) mapToRoute(obj client.Object, routeType core.RouteType) []reconcile.Request {
+	ctx := context.Background()
+	routes := h.mapper.ServiceImportToRoutes(ctx, obj.(*mcs_api.ServiceImport), routeType)
 
+	var requests []reconcile.Request
 	for _, route := range routes {
```

Comment:
nil check on routes or change mapper to return empty lists when nothing to return?

Reply:
I prefer nil slice over empty slice though, we can still iterate over it. https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices

Reply:
nvm, I forgot that nil is also empty list, so range works fine